client/ayon_core/pipeline/editorial.py

```diff
@@ -419,17 +419,50 @@ def get_media_range_with_retimes(otio_clip, handle_start, handle_end):
     else:
         # compute retimed range
         media_in_trimmed = conformed_source_range.start_time.value + offset_in
-        media_out_trimmed = media_in_trimmed + (
-            (
-                conformed_source_range.duration.value
-                * abs(time_scalar)
-                + offset_out
-            ) - 1
-        )
+
+        offset_duration = conformed_source_range.duration.value * abs(time_scalar)
+        offset_duration -= 1  # duration 1 frame -> freeze frame -> end = start + 0
+        offset_duration = max(0, offset_duration)  # negative duration = frozen frame
+        media_out_trimmed = media_in_trimmed + offset_duration
 
         media_in = available_range.start_time.value
         media_out = available_range.end_time_inclusive().value
 
+    if time_warp_nodes:
+        # Naive approach: Resolve consecutive timewarp(s) on range,
+        # then check if plate range has to be extended beyond source range.
+        frame_range = [media_in_trimmed]
+        in_frame = media_in_trimmed + time_scalar
+        while in_frame <= media_out_trimmed:
+            frame_range.append(in_frame)
+            in_frame += time_scalar
+
+        for tw_idx, tw in enumerate(time_warp_nodes):
+            for idx, frame_number in enumerate(frame_range):
+                # First timewarp, apply on media range
+                if tw_idx == 0:
+                    frame_range[idx] = round(frame_number + tw["lookup"][idx] * time_scalar)
+                # Consecutive timewarp, apply on the previous result
+                else:
+                    new_idx = round(idx + tw["lookup"][idx])
+
+                    if not 0 <= new_idx < len(frame_range):
+                        # TODO: implementing this would need to actually have
+                        # retiming engine resolve process within AYON, resolving wraps
+                        # as curves, then projecting those into the previous media_range.
+                        raise NotImplementedError(
+                            "Unsupported consecutive timewarps (out of computed range)"
+                        )
+
+                    frame_range[idx] = frame_range[new_idx]
```

Comment:
Reading `if not 0 <= new_idx < len(frame_range):` is pretty challenging...
```suggestion
                    frame_range[idx] = round(frame_number + tw["lookup"][idx] * time_scalar)
                    continue

                # Consecutive timewarp, apply on the previous result
                new_idx = round(idx + tw["lookup"][idx])
                if 0 <= new_idx < len(frame_range):
                    frame_range[idx] = frame_range[new_idx]
                    continue

                # TODO: implementing this would need to actually have
                # retiming engine resolve process within AYON, resolving wraps
                # as curves, then projecting those into the previous media_range.
                raise NotImplementedError(
                    "Unsupported consecutive timewarps (out of computed range)"
                )

```

Reply:
I've changed it as per your suggestion.